pkg/controllers/infracluster/infracluster_controller.go

```diff
@@ -0,0 +1,288 @@
+	InfraClusterControllerDegradedCondition  = "InfraClusterControllerDegraded"
+
+	defaultCAPINamespace    = "openshift-cluster-api"
+	defaultInfraClusterName = "cluster"
+	clusterOperatorName     = "cluster-api"
+	// This is the managedByAnnotation value that this controller sets by default when it creates an InfraCluster object.
+	// If the managedByAnnotation key is set, and it has this as the value, it means this controller is managing the InfraCluster.
+	managedByAnnotationValueClusterCAPIOperator = ""
+)
+
+type InfraClusterController struct {
+	operatorstatus.ClusterOperatorStatusClient
+	Scheme       *runtime.Scheme
+	Images       map[string]string
+	RestCfg      *rest.Config
+	Platform     configv1.PlatformType
+	Infra        configv1.Infrastructure
+	InfraCluster client.Object
+}
+
+// Reconcile reconciles the cluster-api ClusterOperator object.
+func (r *InfraClusterController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx).WithName("InfraClusterController")
+
+	res, err := r.reconcile(ctx, log)
+	if err != nil {
+		if err := r.setDegradedCondition(ctx, log); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to set conditions for InfraCluster controller: %w", err)
+		}
+		return ctrl.Result{}, fmt.Errorf("error during reconcile: %w", err)
+	}
+
+	if err := r.setAvailableCondition(ctx, log); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to set conditions for InfraCluster controller: %w", err)
+	}
+
+	return res, nil
+}
+
+// reconcile performs the main business logic for installing Cluster API components in the cluster.
+// Notably it fetches CAPI providers "transport" ConfigMap(s) matching the required labels,
+// it extracts from those ConfigMaps the embedded CAPI providers manifests for the components
+// and it applies them to the cluster.
+//
+//nolint:unparam
+func (r *InfraClusterController) reconcile(ctx context.Context, log logr.Logger) (ctrl.Result, error) {
+	switch r.Platform {
+	case configv1.AWSPlatformType:
+		if err := r.ensureAWSCluster(ctx); err != nil {
+			return ctrl.Result{}, fmt.Errorf("error ensuring AWSCluster: %w", err)
+		}
+	case configv1.GCPPlatformType:
+		log.Info(fmt.Sprintf("detected platform %q is supported, but not implemented yet", r.Platform))
+	case configv1.PowerVSPlatformType:
+		log.Info(fmt.Sprintf("detected platform %q is supported, but not implemented yet", r.Platform))
+	case configv1.VSpherePlatformType:
+		log.Info(fmt.Sprintf("detected platform %q is supported, but not implemented yet", r.Platform))
+	case configv1.OpenStackPlatformType:
+		log.Info(fmt.Sprintf("detected platform %q is supported, but not implemented yet", r.Platform))
+	default:
+		log.Info(fmt.Sprintf("detected platform %q is not supported, skipping capi controllers setup", r.Platform))
+	}
+
+	// At this point, the InfraCluster exists.
+	managedByAnnotationVal, found := r.InfraCluster.GetAnnotations()[clusterv1.ManagedByAnnotation]
```

Comment:
I'm surprised that the InfraCluster is a member of the reconciler 🤔

Reply:
Yes, that was to simplify the code but it is not good practice as this could be reused and be stale. We are changing this.